### PR TITLE
Correct `UnitQuaternion` with `Quaternions.jl`

### PR DESCRIPTION
**Pull request by hyrodium**
This PR branch is based on #173, so please merge the PR first.

## Done
* Add dependency on Quaternions.jl
  * https://github.com/JuliaGeometry/Rotations.jl/pull/140#issuecomment-953446907
  * #168
* Remove some incorrect methods
  * `conj(::UnitQuaternion)`
  * `-(::UnitQuaternion)`
  * `*(::Real, ::UnitQuaternion)`
  * `LinearAlgebra.norm(::Real, ::UnitQuaternion)`
  * `LinearAlgebra.normalize(::Real, ::UnitQuaternion)`
  * (and etc.)
* Add type conversion
  * `UnitQuaterinon` <-> `Quaternion`
* Update `Rotations.pure_quaternion`
  * Rename to `_pure_quaternion` because this is "private" method.
  * The return value type is now `Quaternion`. (previously, `UnitQuaternion`)

## WIP, not in this PR
* `UnitQuaternoin{T}` sitll have fields `w::T`, `x::T`, `y::T` and `z::T`, not `q::Quaternion{T}`.
  * I'd like to keep this PR small, so this filed will be changed in another PR.
* The name of unit quaternion type is still `UnitQuaternion`.
  * This will be changed to `QuatRotation` in the future.
* The return value type of `_pure_quaternion` is `Quaternion`.
  * This should be `InfinitesimalQuatRotation <: InfinitesimalRotation` in the future.

### CODE REVIEW

File: src/unitquaternion.jl
Comment (andyferris) [inline src/unitquaternion.jl:231]: We shouldn't be defining things on behalf of Quaternions.jl.

File: src/unitquaternion.jl
Comment (andyferris) [inline src/unitquaternion.jl:230]: The `vecnorm` must be defined in terms of the matrix. Which is orthogonal, so that's easy. Should we just define this instead?

```julia
vecnorm(::Rotation{N, T}) where {N, T} = convert(N, T)
```

File: src/unitquaternion.jl
Comment (andyferris) [inline src/unitquaternion.jl:318]: Aren't these defined for all rotations? (I.e. can they be deleted instead?)

**Review by andyferris — APPROVED** [review]
Nice @hyrodium.

I have no idea whether this is a bugfix or a breaking change. It's not exactly a new feature. It's possible that someone will rely on the current functionality.

File: src/unitquaternion.jl
Comment (hyrodium) [reply src/unitquaternion.jl]: Thanks, that was my mistake. I've removed the line.

File: src/unitquaternion.jl
Comment (hyrodium) [reply src/unitquaternion.jl]: The `vecnorm` function was to calculate the norm of the imaginary part of a quaternion, and it was used only in `_log_as_quat`. So, I've removed the function. (https://github.com/JuliaGeometry/Rotations.jl/pull/175/commits/09e03135986c7e3dcacf7425f2d0e19b94867ba1)

File: src/unitquaternion.jl
Comment (hyrodium) [reply src/unitquaternion.jl]: With original definition:
```julia
julia> @benchmark q\SVector(1,2,3)  # q isa UnitQuaternion
BenchmarkTools.Trial: 10000 samples with 996 evaluations.
 Range (min … max):  25.933 ns …  1.724 μs  ┊ GC (min … max): 0.00% … 97.61%
 Time  (median):     26.546 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   28.727 ns ± 42.051 ns  ┊ GC (mean ± σ):  3.85% ±  2.59%

  ▂▆█▇▄▂                        ▁▂▂▂▂▂▁▁                      ▁
  ██████▇▆▇▆▆▆▅▅▅▅▃▅▅▆▆▆▅▅▅▆▇▇▇██████████▇▆▆▆▅▄▄▆▆▆▅▅▆▅▄▅▄▄▄▆ █
  25.9 ns      Histogram: log(frequency) by time      38.8 ns <

 Memory estimate: 32 bytes, allocs estimate: 1.
```

If Removing the definition:
```julia
julia> @benchmark q\SVector(1,2,3)  # q isa UnitQuaternion
BenchmarkTools.Trial: 10000 samples with 780 evaluations.
 Range (min … max):  160.083 ns …  2.153 μs  ┊ GC (min … max): 0.00% … 92.03%
 Time  (median):     164.683 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   167.640 ns ± 44.395 ns  ┊ GC (mean ± σ):  0.59% ±  2.06%

        ▂▃█▁▁                                                   
  ▂▂▁▁▃▇█████▇▇▇▄▄▃▃▃▃▃▃▃▄▄▄▄▃▃▃▃▃▃▃▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂ ▃
  160 ns          Histogram: frequency by time          186 ns <

 Memory estimate: 32 bytes, allocs estimate: 1.
```

This performance deterioration is because `(\).(::Rotation, ::SVector{3})` is not defined.

Comment (hyrodium) [reply src/unitquaternion.jl]: So, we need to update the `\` definition here. With this update, we'll get more performance with other rotation types.

```julia
julia> using Rotations, StaticArrays, BenchmarkTools

julia> r = RotX(2.3)
3×3 RotX{Float64} with indices SOneTo(3)×SOneTo(3)(2.3):
 1.0   0.0        0.0
 0.0  -0.666276  -0.745705
 0.0   0.745705  -0.666276

julia> @benchmark r\SVector(1,2,3)
BenchmarkTools.Trial: 10000 samples with 861 evaluations.
 Range (min … max):  138.636 ns …  1.605 μs  ┊ GC (min … max): 0.00% … 90.34%
 Time  (median):     154.602 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   156.485 ns ± 31.911 ns  ┊ GC (mean ± σ):  0.45% ±  2.01%

                             ▂▇▇▅█▆▄▃▄                          
  ▂▁▁▁▁▂▁▂▂▁▂▂▂▂▂▂▂▂▃▂▂▃▃▄▄▄▆█████████▇█▄▃▄▃▄▃▄▄▄▄▅▅▅▄▃▃▂▂▂▂▂▂ ▄
  139 ns          Histogram: frequency by time          168 ns <

 Memory estimate: 32 bytes, allocs estimate: 1.

julia> Base.:\(r::Rotation, v::SVector{3}) = inv(r)*v

julia> @benchmark r\SVector(1,2,3)
BenchmarkTools.Trial: 10000 samples with 994 evaluations.
 Range (min … max):  29.926 ns …  1.726 μs  ┊ GC (min … max): 0.00% … 97.98%
 Time  (median):     30.742 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   32.839 ns ± 44.139 ns  ┊ GC (mean ± σ):  3.54% ±  2.59%

  ▃███▇▅▃▂                     ▁▂▂▂▂▂▂▂▁                      ▂
  ██████████▇▇▇▆▇█▆▆▆▆▅▆▇▆▆▇▇▇███████████▇▆▆▅▅▅▆▅▅▆▅▄▄▄▅▅▅▅▅▆ █
  29.9 ns      Histogram: log(frequency) by time      43.2 ns <

 Memory estimate: 32 bytes, allocs estimate: 1.
```

Comment (hyrodium) [reply src/unitquaternion.jl]: I've added the method. (https://github.com/JuliaGeometry/Rotations.jl/pull/175/commits/9647c8476a848199b124dc20807dc956bc9f1102)

**Comment by hyrodium** [comment]
>I have no idea whether this is a bugfix or a breaking change. It's not exactly a new feature. It's possible that someone will rely on the current functionality.

Should we release `v2.0.0`? I'd like to think that was a bug...

**Comment by hyrodium** [comment]
@andyferris 
Can I merge this PR? I've committed after your approval.

**Review by andyferris — APPROVED** [review]
@hyrodium yes feel free to merge

**Comment by dehann** [comment]
Kindly request in future to please add deprecation cycles as per semver for breaking changes.  You can easily overload `Base.get/setproperty` with a deprecation warning.  In our case I had to dig to find out why `quat.w` no longer worked, which was replaced by `quat.q.s` or `.v1/v2/v3`.  Here is a pattern example we use in such cases:
https://github.com/JuliaRobotics/IncrementalInference.jl/wiki/Coding-Templates#deprecate-struct-field

**Comment by hyrodium** [comment]
Sorry for the inconvenience. I thought the fields are just internal parameters because we provide `Rotations.params` method.
This function is not exported, but will be maintained.

I think we need to distinguish among exported method, non-exported method and internal use method.

* exported method
  * e.g.`RotXYZ`, `nearest_rotation`
  * These methods should be maintained.
* non-exported method
  * e.g. `Rotations.params`, `Rotations.jacobian`
  * These methods also should be maintained.
  * The reason for non-exported is just avoiding name conflict.
* internal use method
  * e.g. `_puer_quaternion`
  * These functions are regarded as an internal method, and may be changed without breaking release.
  * In `Rotations.jl`, many functions should be renamed with underscore `_`. (e.g. [`vee`](https://github.com/JuliaGeometry/Rotations.jl/blob/v1.1.0/src/util.jl#L41), [`lmult`](https://github.com/JuliaGeometry/Rotations.jl/blob/v1.1.0/src/unitquaternion.jl#L375))
    * I'm not sure if such renamings would be a breaking change.
    * [See blue guide for more information](https://github.com/invenia/BlueStyle#function-naming).
  * Also sorry, for the inconvenience(https://github.com/RoboticExplorationLab/RobotDynamics.jl/issues/12)

**Comment by bzinberg** [comment]
@dehann, thanks for your insights on this!  The breakage now has its own issue, https://github.com/JuliaGeometry/Rotations.jl/issues/208, let's move discussion there.

**Pull request merged**
